From cce99f3bdcec86511995941dc09cf94706839bd9 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Mon, 12 Oct 2026 02:34:20 +0200
Subject: [PATCH] MFront: Poisson ratio of the stress update algorithm of the
 isotropic DSLs

With `use_stress_update_algorithm : true`, the isotropic DSLs evaluate
the elastic material properties at the beginning of the time step when
they are not constant. The Poisson ratio was converted to `stress`
instead of `real`: with `@UseQt true`, `1 + mfront_nu_bts` did not
compile.

New test law Plasticity8, identical to Plasticity6 except for its
Poisson ratio, which depends formally on the temperature (so that it is
evaluated at the beginning of the time step) but is constant: its test
(plasticity8.mtest) is the one of Plasticity6, with the same results.
---
 .../IsotropicBehaviourCodeGeneratorBase.cxx   |  2 +-
 mfront/tests/behaviours/CMakeLists.txt        |  1 +
 mfront/tests/behaviours/Plasticity8.mfront    | 22 +++++++++++++++++++
 .../tests/behaviours/generic/CMakeLists.txt   |  2 ++
 .../behaviours/generic/plasticity8.mtest      | 13 +++++++++++
 5 files changed, 39 insertions(+), 1 deletion(-)
 create mode 100644 mfront/tests/behaviours/Plasticity8.mfront
 create mode 100644 mfront/tests/behaviours/generic/plasticity8.mtest

diff --git a/mfront/src/IsotropicBehaviourCodeGeneratorBase.cxx b/mfront/src/IsotropicBehaviourCodeGeneratorBase.cxx
index ac52d57f9..c2dd2b00c 100644
--- a/mfront/src/IsotropicBehaviourCodeGeneratorBase.cxx
+++ b/mfront/src/IsotropicBehaviourCodeGeneratorBase.cxx
@@ -230,7 +230,7 @@ namespace mfront {
         }
         if (!this->bd.isMaterialPropertyConstantDuringTheTimeStep(emps[1])) {
           this->writeMaterialPropertyCheckBoundsEvaluation(os, emps[1], bts);
-          os << "const auto mfront_nu_bts = stress(";
+          os << "const auto mfront_nu_bts = real(";
           this->writeMaterialPropertyEvaluation(os, emps[1], bts);
           os << ");\n";
         } else {
diff --git a/mfront/tests/behaviours/CMakeLists.txt b/mfront/tests/behaviours/CMakeLists.txt
index 32425f504..4ffbe5651 100644
--- a/mfront/tests/behaviours/CMakeLists.txt
+++ b/mfront/tests/behaviours/CMakeLists.txt
@@ -148,6 +148,7 @@ install_mfront_data(tests/behaviours Plasticity4.mfront)
 install_mfront_data(tests/behaviours Plasticity5.mfront)
 install_mfront_data(tests/behaviours Plasticity6.mfront)
 install_mfront_data(tests/behaviours Plasticity7.mfront)
+install_mfront_data(tests/behaviours Plasticity8.mfront)
 install_mfront_data(tests/behaviours BerveillerZaouiPolyCrystal.mfront)
 install_mfront_data(tests/behaviours BerveillerZaouiPolyCrystal_DD_CC.mfront)
 install_mfront_data(tests/behaviours BerveillerZaouiPolyCrystal_DD_CFC.mfront)
diff --git a/mfront/tests/behaviours/Plasticity8.mfront b/mfront/tests/behaviours/Plasticity8.mfront
new file mode 100644
index 000000000..757b6ed2f
--- /dev/null
+++ b/mfront/tests/behaviours/Plasticity8.mfront
@@ -0,0 +1,22 @@
+@DSL IsotropicPlasticMisesFlow{use_stress_update_algorithm : true};
+@Behaviour Plasticity8;
+@Author Thomas Helfer;
+@Date 15 / 02 / 2025;
+
+@Epsilon 1e-15;
+@UseQt true;
+
+@Parameter temperature Ta = 2e3;
+@Parameter stress s0 = 1e9;
+s0.setGlossaryName("YieldStrength");
+
+// the Poisson ratio depends formally on the temperature, so that it is
+// evaluated at the beginning of the time step by the stress update
+// algorithm, but is constant: the results are the ones of Plasticity6
+@ElasticMaterialProperties{"150e9*(1-T/Ta)", "0.3+0*T/Ta"};
+
+@FlowRule {
+  f = seq - s0;
+  df_dseq = 1;
+  df_dp = stress{};
+}
diff --git a/mfront/tests/behaviours/generic/CMakeLists.txt b/mfront/tests/behaviours/generic/CMakeLists.txt
index c2ed39862..f1276d75b 100644
--- a/mfront/tests/behaviours/generic/CMakeLists.txt
+++ b/mfront/tests/behaviours/generic/CMakeLists.txt
@@ -190,6 +190,7 @@ if(enable-mfront-quantity-tests)
        Plasticity5
        Plasticity6
        Plasticity7
+       Plasticity8
        PlasticityWithoutAutomaticTemperatureDeclaration
        StrainHardeningCreep3
        StrainHardeningCreep5
@@ -819,6 +820,7 @@ if(enable-mfront-quantity-tests)
   test_generic(plasticity5 plasticity.ref)
   test_generic(plasticity6)
   test_generic(plasticity7 plasticity.ref)
+  test_generic(plasticity8)
   test_generic(plasticity_without_temperature_declaration plasticity.ref)
   test_generic(TensorialExternalStateVariableTest)
 endif(enable-mfront-quantity-tests)
diff --git a/mfront/tests/behaviours/generic/plasticity8.mtest b/mfront/tests/behaviours/generic/plasticity8.mtest
new file mode 100644
index 000000000..f05aefef3
--- /dev/null
+++ b/mfront/tests/behaviours/generic/plasticity8.mtest
@@ -0,0 +1,13 @@
+@Author HELFER Thomas;
+@Date   15/02/2025;
+
+@ModellingHypothesis 'Tridimensional';
+@Behaviour<generic> @library@ 'Plasticity8';
+@ExternalStateVariable<function> 'Temperature' '400 + 50 * sin(t)';
+@Real 'Ta' 2e3;
+@Evolution<function> 'E' '150e9 * (1 - Temperature / Ta)';
+
+@ImposedStrain 'EXX' {0 : 0, 1 : 1e-3};
+@Times {0, 1 in 10};
+
+@Test<function> 'SXX' 'E * EXX' 1e-2;
\ No newline at end of file
-- 
2.43.0

